src/components/view-3d/marker.tsx

```diff
@@ -0,0 +1,120 @@
+import React, { useEffect, useRef } from "react";
+import * as THREE from "three";
+import { useThree } from "../../react-three-hook";
+import { observer } from "mobx-react";
+import { useStores } from "../../use-stores";
+import { IThreeContext } from "../../react-three-hook/threejs-manager";
+import { ftToViewUnit, PLANE_WIDTH } from "./helpers";
+import { useInteractions } from "./use-interactions";
+import { SpriteMaterial } from "three";
+
+interface IProps {
+  markerImg: string;
+  markerHighlightImg: string;
+  position: {x: number, y: number};
+  setPosition: (x: number, y: number) => void;
+  getTerrain: () => THREE.Mesh;
+  lockOnSimStart?: boolean;
+}
+
+const SIZE = 0.06 * PLANE_WIDTH;
+
+const setupMesh = ({ scene }: IThreeContext) => {
+  const sprite = new THREE.Sprite();
+  // Move anchor point to bottom.
+  sprite.center.y = 0;
+  sprite.scale.set(SIZE, SIZE, 1);
+  // Ensure that sprite is always rendered on top of other geometry, so e.g. it doesn't disappear under a mountain.
+  sprite.renderOrder = 1;
+  scene.add(sprite);
+  return sprite;
+};
+
+const getMaterial = (imgSrc: string) => {
+  const image = document.createElement("img");
+  image.src = imgSrc;
+  image.onload = () =>  { texture.needsUpdate = true; };
+  const texture = new THREE.Texture(image);
+  const material = new THREE.SpriteMaterial({ map: texture });
+  material.depthTest = false;
+  return material;
+};
+
+export const Marker: React.FC<IProps> = observer(({
+  markerImg, markerHighlightImg, position, setPosition, getTerrain, lockOnSimStart = false
+}) => {
+  const defMaterial = useRef<SpriteMaterial>();
+  const highlightMaterial = useRef<SpriteMaterial>();
+  const { simulation, ui } = useStores();
+
+  const { getEntity } = useThree<THREE.Sprite>(setupMesh);
+
+  useEffect(() => {
+    defMaterial.current = getMaterial(markerImg);
+    highlightMaterial.current = getMaterial(markerHighlightImg);
+    const marker = getEntity();
+    if (marker) {
+      marker.material = defMaterial.current;
+    }
+    return () => {
+      defMaterial!.current!.map!.dispose();
+      defMaterial!.current!.dispose();
+      highlightMaterial!.current!.map!.dispose();
+      highlightMaterial!.current!.dispose();
+    };
```

Comment:
Seems like a lot! of! not-null!

Reply:
Yeap, my biggest problem with hooks and refs is actually that everything can be `null` and there're so many `if`s required (or `!`).

Reply:
In TypeScript 3.7, you can use optional chaining (`?.`) for the best of both worlds -- test and proceed if present.

Reply:
Good to know, thanks!